MWI: Bound Keypair Joining - Minimal Client Implementation

timothyb89 opened the pull request:
This contains a minimal client-side implementation of bound keypair joining for `tbot`. This first iteration supports only preregistered keys, and so adds a new `tbot keypair create` subcommand that generates a keypair and prints a public key that can be copied into a token.

In an effort to still take advantange of tbot's secure filesystem abstraction and to not permanently tie the new join method to it, this includes a minimal filesystem wrapper to adapt tbot's destinations to a generic file read/write interface.

Protos PR: https://github.com/gravitational/teleport/pull/53566
Backend PR: https://github.com/gravitational/teleport/pull/54371
RFD: https://github.com/gravitational/teleport/pull/52546

changelog: Add experimental support for Bound Keypair joining

timothyb89 committed:
MWI: Bound Keypair Joining - Minimal Client Implementation

This contains a minimal client-side implementation of bound keypair
joining for `tbot`. This first iteration supports only preregistered
keys, and so adds a new `tbot keypair create` subcommand that
generates a keypair and prints a public key that can be copied into
a token.

In an effort to still take advantange of tbot's secure filesystem
abstraction and to not permanently tie the new join method to it,
this includes a minimal filesystem wrapper to adapt tbot's
destinations to a generic file read/write interface.

timothyb89 committed:
Rebase join client to handle interactive rotation messages

This updates the client implementation to handle the revised challenge
handler signature. Rotation still isn't implemented (see TODOs), but
the message handler now explicitly accounts for rotation-type requests
from the server.

timothyb89 committed:
Fix breakage after upstream changes, idempotent key generation

The `keypair generate` command is now idempotent and will not generate
a new keypair if one already exists in the specified destination.

Also, added a `--format=json` flag to print the key as
machine-readable JSON.

timothyb89 committed:
Remove unnecessary whitespace

timothyb89 committed:
Small TODO and doc comment updates

strideynet commented on lib/tbot/cli/keypair_create.go:
```diff
@@ -0,0 +1,47 @@
+	Format      string
+}
+
+// NewKeypairCreateCommand initializes the `keypair create` command and returns
```
I'd be tempted to use a more semantic command than "create" here - but nothing immediately comes to mind. I'll mull it over.

timothyb89 replied:
Possible alternatives:
- `tbot keypair generate`
- `tbot keypair new`
- other suggestions?

I have no strong feelings on the UX here, so let me know what you think!

strideynet replied:
Yeah - perhaps "generate" is nicer. My only other worry is that "keypair" alone may not indicate sufficiently that this is something to do with bound keypair joining - and `tbot` does a lot with keypairs!

timothyb89 replied:
Per discussion in our 1x1 - I've hidden this for now but left it otherwise unchanged. We may still opt to change the name here before the feature is released.

timothyb89 committed:
Fix lints

timothyb89 commented on lib/tbot/cli/keypair_create.go:
```diff
@@ -0,0 +1,47 @@
+// NewKeypairCreateCommand initializes the `keypair create` command and returns
+// a struct to contain the parse result.
+func NewKeypairCreateCommand(parentCmd KingpinClause, action func(*KeypairCreateCommand) error) *KeypairCreateCommand {
+	cmd := parentCmd.Command("create", "Create a keypair to preregister for bound-keypair joining")
```
TODO: hide the command for now so we don't commit to a CLI contract

timothyb89 committed:
Mark `keypair create` as a hidden command for now

This hides it from `tbot help` so we can opt to tweak the CLI contract
before fully releasing the feature. It will be unhidden in the near
future.

timothyb89 commented:
> @timothyb89 - this PR will require admin approval to merge due to its size. Consider breaking it up into a series smaller changes.

Rebase issue, please ignore 😅

timothyb89 committed:
Fix typo
